zmq-multiplex.slide

```diff
@@ -0,0 +1,136 @@
+ZMQ multiplex between golang coroutine and python thread
+10 June 2017
+
+Rick Mak
+Oursky
+rickmak@oursky.com
+@rickmak
+CC81327D
+
+* What we want to do
+
+- Enable building of scalable software with less learning curve
+- Provide foundation for developer to focus business requirement and UI
+- Save developer from deployment hassle
+
+* Skygear-server
+
+- Design for web development
+- Written in golang
+- Provide basic function like Auth, Record CRUD, PubSub via core
+- Provide plugin for extending function, fulfilling business requirement.
+
+* ZMQ
+
+- High performance messaging queue
+- No message broker
+
+* Python
+
+- Imperative
+- Easy to write
+- Easy to read
+- Code quickly
+
+Actually we have JS too.
+
+* Data flow
+
+.image zmq-multiplex/dataflow.png
+
+- Skygear-server is written in golang and able to handle thousands of requests
+- For request triggering cloud code, a message will send to worker via zmq
+
+* Wire Protocol
+
+    [WORKER-ADDR|0|body....]
+
+- Simple
+- Routing based on WORKER-ADDR
+
+The protocol also have heartbeat, details:
+https://rfc.zeromq.org/spec:6/PPP
+
+* Problem
+
+- Python-side resource exhausted on high load
+- Thread overhead is much higher than goroutine
+- Recursion, cloud code will call back skygear-server
+- User don't want to handle the resource problem
+- Providing a “just works” solution
+
+* Solution
+
+- Reuse of thread awaiting skygear-server response
+- Provide configurable dynamic Thread pool
+- Limit the recursion to a configurable depth
+
+* Wire Protocol
+
+Extending the old protocol to support multiplex.
+
+- [WORKER-ADDR|0|REQ|BOUNCE-COUNT|REQ-ID|0|body....]
+
+1. Worker address
+   Same as before, for routing the message to the correct worker
+
+2. Message type
+3. Bounce Count
+4. Request ID
+   For worker reuse, and route recursion call to the same worker
+
+5. Message body
+   Actual message body
+
+* Example
+
+Plugin code:
+
+    @op('foo:hello')
```

Comment:
I think this example should move way earlier in the talk because it helps the audience to visualize what cloud code looks like and how it works.